galley/pkg/config/analysis/analyzers/auth/mtls.go

```diff
@@ -0,0 +1,355 @@
+
+			if svcSelector.Empty() || !svcSelector.Matches(podLabels) {
+				return true
+			}
+
+			// This pod is selected for this service - ensure there's a sidecar.
+			foundMatchingPods = true
+			sidecarFound := false
+			for _, container := range pod.Spec.Containers {
+				if container.Name == "istio-proxy" {
+					sidecarFound = true
+				}
+			}
+
+			if !sidecarFound {
+				fqdnsWithoutSidecars[fqdn] = struct{}{}
+			}
+			return true
+		})
+
+		if !foundMatchingPods {
+			fqdnsWithoutSidecars[fqdn] = struct{}{}
+		}
+
+		return true
+	})
+
+	// While we visit every item, collect the set of namespaces that exist. Note
+	// that we will collect the namespace name for all resource types - this
+	// ensures our analyzer still behaves correctly even if namespaces are
+	// implicitly defined.
+	namespaces := make(map[string]struct{})
+
+	c.ForEach(metadata.K8SCoreV1Namespaces, func(r *resource.Entry) bool {
+		_, name := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[name] = struct{}{}
+		return true
+	})
+
+	pc := mtls.NewPolicyChecker(fqdnToNameToPort)
+	c.ForEach(metadata.IstioAuthenticationV1Alpha1Meshpolicies, func(r *resource.Entry) bool {
+		pc.AddMeshPolicy(r, r.Item.(*v1alpha1.Policy))
+		return true
+	})
+
+	c.ForEach(metadata.IstioAuthenticationV1Alpha1Policies, func(r *resource.Entry) bool {
+		ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[ns] = struct{}{}
+
+		err := pc.AddPolicy(r, r.Item.(*v1alpha1.Policy))
+		if err != nil {
+			// AddPolicy can return a NamedPortInPolicyNotFoundError - if it
+			// does we can print a useful message.
+			// TODO this should be in its own analyzer, and ignored here.
+			if missingPortNameErr, ok := err.(mtls.NamedPortInPolicyNotFoundError); ok {
+				c.Report(metadata.IstioAuthenticationV1Alpha1Meshpolicies,
+					msg.NewPolicySpecifiesPortNameThatDoesntExist(r, missingPortNameErr.PortName, missingPortNameErr.FQDN))
+				return true
+			} else {
+				c.Report(metadata.IstioAuthenticationV1Alpha1Meshpolicies, msg.NewInternalError(r, err.Error()))
+				return false
+			}
+		}
+		return true
+	})
+
+	drc := mtls.NewDestinationRuleChecker(rootNamespace)
+	c.ForEach(metadata.IstioNetworkingV1Alpha3Destinationrules, func(r *resource.Entry) bool {
+		ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[ns] = struct{}{}
+
+		drc.AddDestinationRule(r, r.Item.(*v1alpha3.DestinationRule))
+		return true
+	})
+
+	// Here we explicitly handle the common case where a user specifies a
+	// MeshPolicy with no global DestinationRule. We also track if we report a
+	// problem with the global configuration. This is used later to suppress
+	// reporting a message for every service/namespace combination due to the
+	// same misconfiguration.
+	globalMTLSMisconfigured := false
+	mpr := pc.MeshPolicy()
+	globalMtls, globalDR := drc.DoesNamespaceUseMTLSToService(rootNamespace, rootNamespace, mtls.NewTargetService("*.svc.cluster.local"))
```

Comment:
FYI, there is no formal definition for global DR. Thus any thing that can match to all services in the mesh is considered as global DR. Thus, `*`, `*.local`, `*.cluster.local` etc are all can be global DR. `*.svc.cluster.local` could be the one we use in our installation.

Reply:
Right, the code actually handles that case. If the target service has a wildcard in it, it looks for matching destination rules that are a strict superset - this means that as long as the user has a rule that's equal to or more general than `*.svc.cluster.local` it will match.

Reply:
That `.svc.cluster.local` fragment should probably be a shared constant between analyzers, I know we've got more than one thing looking for it.

Reply:
Went ahead and made it a shared constant.